juju_verify/verifiers/__init__.py

```diff
@@ -41,38 +42,36 @@
 }
 
 
-def get_verifier(units: List[Unit]) -> BaseVerifier:
+def get_verifiers(units: List[Unit]) -> Iterator[BaseVerifier]:
     """Implement Factory function "verifier" creator for the supplied units.
 
     :param units: Juju unit(s) for which you want to produce verifier
     :return: Correct verifier for given unit(s)
     :raises CharmException: Raised if units do not belong to the same charm or
         if the charm is unsupported for verification
     """
-    charm_types = set()
-
     if not units:
         raise CharmException("List of units can not be empty when creating verifier")
+
+    charms = defaultdict(list)
+
     for unit in units:
         charm_type = parse_charm_name(unit.data.get("charm-url", ""))
         logger.debug("Inferred charm for unit %s: %s", unit.entity_id, charm_type)
-        charm_types.add(charm_type)
-
-    if len(charm_types) > 1:
-        raise CharmException(
-            f"Units are not running same charm. Detected types: {charm_types}"
-        )
-
-    charm = charm_types.pop()
+        charms[charm_type].append(unit)
 
-    verifier = SUPPORTED_CHARMS.get(charm)
+    for charm, charm_units in charms.items():
+        logger.info("===[%s]===", ", ".join([unit.entity_id for unit in charm_units]))
+        verifier = SUPPORTED_CHARMS.get(charm)
+        if verifier is None:
+            supported_charms = os.linesep.join(SUPPORTED_CHARMS.keys())
+            logger.error(
+                "Charm '%s' is not supported by juju-verify. Supported charms:%s%s",
+                charm,
+                os.linesep,
+                supported_charms,
+            )
+            continue
```

Comment:
The previous workflow decided if all units being verified were using the same charm. The options were: a) verify all units using that single verifier b) or raise an error because the charm used is not supported.

Now, any units with supported charms (no matter they are different) will be tested. This opens the door to: a) test some units, b) raise an error for unsupported charms, c) test some more units with supported charms.

The proposed approach from martin is to remove (c), but (a) could still happen before (b). If we want to keep the same type of output as before, All units should be grouped by used charms, and raise the error if any of those charms is not supported (so (b) should go before (a) so there are no checks run at all if unsupported charms are backing some of the units that a user wants to verify).

Under the above premises, do we still want to "break" rather than "continue"?

Reply:
> The proposed approach from martin is to remove (c), but (a) could still happen before (b)

When I said "stop execution" in my original comment, I meant completely stop (e.g. raise exception like it was implemented originally), not just `break` the loop.
At this stage we are only gathering the verifiers, no checks are performed yet, so the scenario "Some supported checks are executed and then we raise exception on unsupported charms" is not possible here.

Reply:
Imo the approach now is better than raising any error, because like this I could run check with multiple units from different charms and still get nicely grouped results.

An example:
```bash
$ juju-verify reboot --unit ceph-osd/0 grafana/0 ceph-mon/0
===[ceph-osd/0]===
Checks:
[OK] check_affected_machines check passed
[WARN] ceph-osd/0 has units running on child machines: ceph-mon/0*
[FAIL] ceph-mon/1: Ceph cluster is unhealthy
[OK] Minimum replica number check passed.
[OK] Availability zone check passed.

Result: Failed

===[grafana/0]===
Charm 'grafana' is not supported by juju-verify. Supported charms:
nova-compute
ceph-osd
ceph-mon
neutron-gateway
===[ceph-mon/0]===
Checks:
[OK] check_affected_machines check passed
[OK] check_has_sub_machines check passed
[OK] Minimum juju version check passed.
[OK] Ceph-mon quorum check passed.
[FAIL] ceph-mon/0: Ceph cluster is unhealthy

Result: Failed
```

Reply:
Thanks for the clarification, Martin.

I understand Martin's concern. We are trying to close 2 bugs in one. However, do you (Martin) still see value on raising the exception instead of making the user aware that such check is not possible (as part of a list of checks per type of unit)? Robert's output for the `grafana/0` unit, as part of other checks looks useful.

Reply:
hmm, you may be right. The original behavior probably made sense only when we could verify just a single charm. The output example from @r-f-g looks fine.